src/runtime_src/core/edge/user/aie/AIEResources.h

```diff
@@ -0,0 +1,200 @@
+    template<size_t NUM_PERF_COUNTERS> class PerformanceCounter
+    {
+    public:
+        int requestPerformanceCounter(short handle);
+        bool requestPerformanceCounter(short handle, size_t index);	
+        void releasePerformanceCounter(short handle, size_t index);
+        int getNumUsedPerformanceCounters();
+
+    protected:
+        PerformanceCounter();
+
+    protected:
+        short performanceCounters[NUM_PERF_COUNTERS];
+        unsigned short numUsedPerformnceCounters;
+    };
+
+    class CoreAndPLModule : public Module
+    {
+    public:
+        int requestStreamEventPort(short handle);
+        void releaseStreamEventPort(short handle, size_t index);
+        int getNumUsedStreamEventPorts();
+    protected:
+        CoreAndPLModule();
+
+    protected:
+        short streamSwitchEventPorts[NUM_STREAM_SWITCH_EVENT_PORTS];
+	unsigned short numUsedStreamSwitchEventPorts;
+    };
+    
+    class CoreModule : public CoreAndPLModule, public PerformanceCounter<NUM_PERF_COUNTERS_CORE>
+    {
+    public:
+        CoreModule();
+
+        int requestProgramCounter(short handle);
+        std::pair<int, int> requestProgramCountersForRange(short handle);
+        void releaseProgramCounter(short handle, size_t index);
+        short getTraceUnitPacketType() override;
+
+    private:
+        short programCounters[NUM_PROGRAM_COUNTERS];
+    };
+    
+    class MemoryModule : public Module, public PerformanceCounter<NUM_PERF_COUNTERS_MEM>
+    {
+    public:
+        short getTraceUnitPacketType() override;
+    };
+    
+    class PLModule : public CoreAndPLModule, public PerformanceCounter<NUM_PERF_COUNTERS_PL>
+    {
+    public:
+        short getTraceUnitPacketType() override;
+    };
+
+    class ShimTile
+    {
+    public:
+        PLModule plModule;
+    };
+    
+    class AIETile
+    {
+    public:
+        CoreModule coreModule;
+        MemoryModule memoryModule;
+    };
+    
+    class AIE
+    {
+    public:
+        /// @param numAIERows Total number of AIE tile rows, excluding shim
+        static void initialize(size_t numColumns, size_t numAIERows);
+        /// @param row Row 0 means the first AIE array row, does not include shim
+        static AIETile* getAIETile(size_t column, size_t row);
+        static ShimTile* getShimTile(size_t column);
+
+        /// Reserve the 0th performance counter for ECC Scrubbing in all core modules
+        static bool reservePerformanceCounterEccScrubbing();
+        
+    private:
+        static AIETile* s_meTiles;
+        static ShimTile* s_shimTiles;
+        static size_t s_numColumns;
+        static size_t s_numAIERows;
+    };
+    
+    enum module_type
+    {
+        core_module,
+        memory_module,
+        pl_module
+    };
+    
+    enum resource_type
+    {
+        performance_counter,
+        trace_event,
+        stream_switch_event_port,
+        event_broadcast,
+        program_counter,
+        group_event,
+        combo_event
+    };
```

Comment:
Can these be enum classes instead of plain enums?
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Renum-class